vsm

```diff
@@ -147,31 +149,55 @@ def process(state, signal, value):
     state.got_signal(signal, value)
 
 
-def run(state):
+def receive():
     '''
-        Loop to grab logic from stdin, then spawn a threads for processing
+        Loop to grab logic from stdin
     '''
     for line in sys.stdin:
         line = line.replace(" ", "").strip('\n')
         if line == "":
             # Ignore blank lines
             pass
         if line == "quit":
-            break
+            exit(0)
         elif "=" in line:
             line = line.split("=")
-            process(state, line[0], line[1])
+            return (line[0], line[1])
         else:
-            process(state, line, None)
-
+            return (line, None)
+
+def run(state, no_loop):
+    try:
+        while True:
+            message = receive()
+            if message == None:
+                exit(0)
+            signal, value = message
+            process(state, signal, value)
+            if no_loop:
```

Comment:
This will not do as some tests send more than one message. Either make 'quit' quit the program for ipc modules as well or - even better - kill the process in the tests after sending/receiving.